client/src/main/java/com/aws/greengrass/cli/adapter/impl/NucleusAdapterIpcClientImpl.java

```diff
@@ -209,6 +225,198 @@ public CreateDebugPasswordResponse createDebugPassword() {
         }
     }
 
+    @Override
+    public void publishToTopic(String topicName, String message) {
+        try {
+            PublishToTopicRequest publishToTopicRequest = new PublishToTopicRequest();
+            PublishMessage publishMessage = new PublishMessage();
+            BinaryMessage binaryMessage = new BinaryMessage();
+            binaryMessage.setMessage(message.getBytes(StandardCharsets.UTF_8));
+            publishMessage.setBinaryMessage(binaryMessage);
+            publishToTopicRequest.setPublishMessage(publishMessage);
+            publishToTopicRequest.setTopic(topicName);
+            getIpcClient().publishToTopic(publishToTopicRequest, Optional.empty());
+        } finally {
+            close();
+        }
+    }
+
+    @Override
+    public void publishToIoTCore(String topicName, String message, String qos) {
+        try {
+            QOS qos1 = QOS.get(qos);
+            PublishToIoTCoreRequest publishToIoTCoreRequest = new PublishToIoTCoreRequest();
+            publishToIoTCoreRequest.setTopicName(topicName);
+            publishToIoTCoreRequest.setPayload(message.getBytes(StandardCharsets.UTF_8));
+            publishToIoTCoreRequest.setQos(qos1);
+            getIpcClient().publishToIoTCore(publishToIoTCoreRequest, Optional.empty());
+        } finally {
+            close();
+        }
+    }
+
+    @Override
+    public void subscribeToTopic(String topic) {
+        try {
+            StreamResponseHandler<SubscriptionResponseMessage> streamResponseHandler =
+                    new SubscriptionResponseHandler(topic);
+            SubscribeToTopicResponseHandler responseHandler =
+                    subscribeToTopic(getIpcClient(), topic, streamResponseHandler);
+            CompletableFuture<SubscribeToTopicResponse> futureResponse =
+                    responseHandler.getResponse();
+            try {
+                futureResponse.get(DEFAULT_TIMEOUT_IN_SEC, TimeUnit.SECONDS);
+                System.out.println("Successfully subscribed to topic: " + topic);
+            } catch (TimeoutException e) {
+                System.err.println("Timeout occurred while subscribing to topic: " + topic);
+            } catch (ExecutionException e) {
+                if (e.getCause() instanceof UnauthorizedError) {
+                    System.err.println("Unauthorized error while publishing to topic: " + topic);
+                } else {
+                    throw e;
+                }
+            }
+
+            // Keep the main thread alive, or the process will exit.
```

Comment:
If I use ctrl+c key combination to kill the CLI, does this exit correctly?

Reply:
Yes, it can exit correctly.